libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2980,6 +2953,81 @@ string YulUtilFunctions::conversionFunction(Type const& _from, Type const& _to)
 	});
 }
 
+string YulUtilFunctions::arrayConversionFunction(ArrayType const& _from, ArrayType const& _to)
+{
+	string functionName =
+		"convert_array_" +
+		_from.identifier() +
+		"_to_" +
+		_to.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		Whiskers templ(R"(
+			function <functionName>(value<?fromCalldataDynamic>, length</fromCalldataDynamic>) -> converted {
+				<?sameTypes>
+					converted := value
+				<!sameTypes>
+					<?toMemory>
+						// Copy the array to a free position in memory, unless it is already in memory.
+						<?fromMemory>
+							converted := value
+						</fromMemory>
+						<?fromStorage>
+							converted := <arrayStorageToMem>(value)
+						</fromStorage>
+						<?fromCalldata>
+							converted := <allocateMemoryArray>(length)
+							<copyToMemory>(value, add(converted, 0x20), length)
+						</fromCalldata>
+					</toMemory>
+
+					<?toStorage>
+						converted := value
```

Comment:
I'm a bit uneasy about seeing the assertions only in the code below. Would it be easier to have some c++ code that detect if there is nothing to do and then return the identity function and only in the more complicated case actually do something?

Reply:
I moved now assertions to the beginning of a function. Also did a bit of refactoring on how body of a function is built.
Hope it looks better now